language/src/regent/optimize_index_launches.t

```diff
@@ -1154,18 +1163,127 @@ local function optimize_loop_body(cx, node, report_pass, report_fail)
     end
   end
 
-  report_pass("loop optimization succeeded")
+  if needs_dynamic_check then
+    report_pass("static loop optimization failed, emitting dynamic check")
+  else
+    report_pass("loop optimization succeeded")
+  end
+
   return {
     preamble = preamble,
     call = call,
     reduce_lhs = reduce_lhs,
     reduce_op = reduce_op,
     args_provably = args_provably,
     free_variables = free_vars,
-    loop_variables = loop_vars
+    loop_variables = loop_vars,
+    needs_dynamic_check = needs_dynamic_check
   }
 end
 
+function init_bitmask_false(bitmask, volume)
+  local stats = terralib.newlist()
+
+  local idx = std.newsymbol(int32, "i")
+  local bitmask_i = util.mk_expr_index_access(util.mk_expr_id_rawref(bitmask), util.mk_expr_id(idx), std.rawref(&bool))
+  local assign = util.mk_stat_assignment(bitmask_i, util.mk_expr_constant(false, bool))
+  stats:insert(assign)
+
+  local values = terralib.newlist()
+  values:insert(util.mk_expr_constant(0, int32))
+  values:insert(util.mk_expr_id(volume))
+
+  return util.mk_stat_for_num(idx, values, util.mk_block(stats))
+end
+
+function get_check_stats(preamble, value, index_expr, bitmask, conflict, pos, loop_var, volume)
+  local stats = terralib.newlist()
+
+  -- Compute value = index_expr(i)
+  preamble:map(function(stat) stats:insert(stat) end)
+  stats:insert(util.mk_stat_assignment(util.mk_expr_id_rawref(value), index_expr))
+
+  local then_block = terralib.newlist()
+
+  local bitmask_value = util.mk_expr_index_access(util.mk_expr_id(bitmask), util.mk_expr_id(value), std.rawref(&bool))
+  local conflict_assign = util.mk_stat_assignment(util.mk_expr_id_rawref(conflict), bitmask_value)
+  then_block:insert(conflict_assign)
+
+  local bitmask_assign_true = util.mk_stat_assignment(bitmask_value, util.mk_expr_constant(true, bool)) 
+  then_block:insert(bitmask_assign_true)
+
+  local save_pos = util.mk_stat_assignment(util.mk_expr_id_rawref(pos), util.mk_expr_id(loop_var))
+  local check_conflict = util.mk_stat_if(util.mk_expr_id(conflict), terralib.newlist { save_pos, util.mk_stat_break() })
+  then_block:insert(check_conflict)
+
+  local cond = util.mk_expr_binary("<", util.mk_expr_id(value), util.mk_expr_id(volume))
+  local bounds_check = util.mk_stat_if(cond, then_block)
+  stats:insert(bounds_check)
+
+  return stats
+end
+
+function insert_dynamic_check(index_launch_ast, unoptimized_loop_ast)
+  local stats = terralib.newlist()
+
+  -- Optimize only if we are indexing into a partition
+  local proceed, p_symbol = pcall(function() return index_launch_ast.call.args[1].value.value end)
```

Comment:
This looks like a hard-coded check for a specific argument?

Also, if this is truly a requirement on the dynamic optimization, it should probably be lifted into the checking phase of the optimization so that we can issue an error message if the loop cannot be dynamically optimized. (If it's just an expedience to make the initial code faster to write, let's generalize it instead.)

Also since we're not doing things with "opaque" objects (like Terra ASTs) here, I'd be inclined to write out the check rather than use `pcall`. I generally reserve `pcall` for cases where such a check *cannot* reasonably be written out (e.g., like I said, when checking whether a Terra AST type checks, which we can't really do by any means other than invoking Terra to compile it).

Reply:
Ok, I've replaced this with a somewhat long check that uses the compiler's `std`. I don't think we can put this anywhere else because we only generate the check for partitions and the task might have other, non-partition arguments which we need to skip.